### PR TITLE
FinalInternalClassFixer - Attribute support

### PR DESCRIPTION
**Pull request by SpacePossum**
closes https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/5782

### CODE REVIEW

**Comment by keradus** [comment]
#5941 got merged, I let myself rebase the PR

File: doc/rules/class_notation/final_internal_class.rst [hunk 1]
```diff
@@ -40,6 +40,16 @@ Allowed types: ``bool``
...
 
+``attribute_exclude``
+~~~~~~~~~~~~~~~~~~~~~
+
+Class level attributes that must be omitted to fix the class, even if all of the
+white list ones are used as well. (case insensitive)
+
+Allowed types: ``array``
+
+Default value: ``['Entity', 'ORM\\Entity', 'ORM\\Mapping\\Entity', 'Mapping\\Entity']``
```
Comment (kubawerlos) [inline doc/rules/class_notation/final_internal_class.rst:51]: Why is this list missing `Document` and `ODM\\Document` that are in `annotation_exclude`?

Comment (SpacePossum) [reply doc/rules/class_notation/final_internal_class.rst:51]: The scope of scope the PR is the linked issue, there is no mention of `Document`. I didn't find it in the doctrine documentation either. We should keep PR's small, so I won't add features that were not requested, you can PR yourself if you want.

Comment (kubawerlos) [reply doc/rules/class_notation/final_internal_class.rst:51]: The scope of this PR is attribute support, and support for mentioned classes is already [merged](https://github.com/doctrine/mongodb-odm/pull/2344) and will go live in next release, I don't understand why they can be added in this PR?

> We should keep PR's small

Absolutely, definitely, that's why I want https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5900 to be merged and improved later step-by-step and what's why I had objection with https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5934 which was too big to decent review as [few](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5962) places were missed.

Comment (keradus) [reply doc/rules/class_notation/final_internal_class.rst:51]:
```suggestion
Default value: ``['Entity', 'ORM\\Entity', 'ORM\\Mapping\\Entity', 'Mapping\\Entity',m 'Document', 'ODM\\Document']``
```
as we use in other fixer - `final_internal_class`

Comment (SpacePossum) [reply doc/rules/class_notation/final_internal_class.rst:51]: >The scope of this PR is attribute support, 

no, it is my PR and I set the scope, go do something else than spamming my PRs

File: src/Fixer/ClassNotation/FinalInternalClassFixer.php [hunk 2]
```diff
@@ -174,6 +179,25 @@ protected function createConfigurationDefinition(): FixerConfigurationResolverIn
                 ->setAllowedTypes(['bool'])
                 ->setDefault(false)
                 ->getOption(),
+            (new FixerOptionBuilder('attribute_exclude', 'Class level attributes that must be omitted to fix the class, even if all of the white list ones are used as well. (case insensitive)'))
```
Comment (keradus) [inline src/Fixer/ClassNotation/FinalInternalClassFixer.php:182]: I started to wonder - what's the need of having `annotation_exclude` and `attribute_exclude` both?
Maybe we should simply have `exclude` (alias for `annotation_exclude`) that is considered for both, annotations and attributes?

 It's like PHP introduced attributes and now ppl are migrating from annotations to use attributes, but the purpose and reasoning is the same, so maybe we don't need to distinguish them on config level?

File: tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php [hunk 3]
```diff
@@ -304,4 +306,77 @@ public function testConfigureSameAnnotationInBothLists(): void
...
+     */
+    public function testAttributeExclude(string $expected, string $input, array $config): void
+    {
+        $config['consider_absent_docblock_as_internal_class'] = true;
+
+        $this->fixer->configure($config);
+        $this->doTest($expected, $input);
+    }
+
+    public function provideAttributeExcludeCases(): \Generator
```
Comment (GrahamCampbell) [inline tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:322]: What happens if we run these tests on PHP 7?

Comment (SpacePossum) [reply tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:322]: The attributes will be tokenized as comments so the fixer won't make changes. We could on > PHP8 already check comments and if these match the attribute-format fix accordingly, would it be worth it?

File: src/Fixer/ClassNotation/FinalInternalClassFixer.php (hunk 2 above)
Comment (SpacePossum) [reply src/Fixer/ClassNotation/FinalInternalClassFixer.php:182]: nice idea, not sure how to do it in a BC way just yet, need to think on it some more

File: tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php (hunk 3 above)
Comment (GrahamCampbell) [reply tests/Fixer/ClassNotation/FinalInternalClassFixerTest.php:322]: maybe, but out of scope of this PR i guess :P

**Pull request closed without merging**